src/compilerOptions/typescriptReferences.ts

```diff
@@ -0,0 +1,222 @@
+        const absPath = ensureAbsolutePath(rawPath, tsConfigDir)
+        recurseTsReference(tsHost, absPath, inputsByOutput, anticycle);
+    }
+    return inputsByOutput;
+}
+
+// group an out-to-in mapping by package (accounting for nested pacakges)
+export function groupByPackage(mappings: ByOutputMap): ByPackageMap {
+    // find all the relevant package roots (bases)
+    // we need to know what node packages the typescript out/in files are in
+    // so that when we get an output, even if it is a PnP virtual path, we find the right out/in pair
+    const byPackage = new Map<string, ByOutputMap>();
+    for (const [output, input] of mappings) {
+        for (const { base, rel: outputRel } of packageAncestorsOf(output)) {
+            const map = byPackage.get(base) || new Map<string, string>();
+            if (map.size == 0) {
+                byPackage.set(base, map);
+            }
+            const inputRel = relative(base, input);
+            map.set(outputRel, inputRel);
+        }
+    }
+    return byPackage;
+}
+
+// Create a resolver that can resolve project subpaths
+// mapping from outputs (e.g. dist/file.js) to their inputs (e.g. src/file.ts)
+export function createTsTargetResolver(references: ITsConfigReference[], tsConfigDir: string): TargetResolver | undefined {
+    const mappings = groupByPackage(buildMappings(references, tsConfigDir));
+    if (mappings.size === 0)
+        return undefined;
+    return mappingsToResolver(mappings);
+}
+
+export function mappingsToResolver(mappings: ByPackageMap) {
+    // take a base and a relative target and return the input, if any, that generates it
+    function tsMapToInput(pkgRoot, subPath) {
+        const inByOutRel = mappings.get(pkgRoot);
+        if (inByOutRel) {
+            const input = inByOutRel.get(subPath);
+            if (input) {
+                return input
+            }
+        }
+        return subPath;
+    }
+
+    // create a SubPathResolver that checks the Out->In maps
+    const tsResolver: SubPathResolver = (base, target, type, props) => {
+        const basicResult = resolveIfExists(base, tsMapToInput(base, target), type, props);
+        if (basicResult) {
+            return basicResult;
+        }
+        // Handle PnP virtual paths
+        if (resolveVirtual) {
+            const realBase = resolveVirtual(base);
+            if (realBase && realBase !== base) {
+                // if we are dealing with virtual paths
+                // we need to send the "real" base path to the ts mapper
+                // but send the original virtual base path as our result
+                var result = resolveIfExists(base, tsMapToInput(realBase, target), type, props);
+                if (result) {
+                    return result;
+                }
+            }
+        }
+        return undefined;
+    }
+
+    // Construct a TargetResolver but inject our own TypeScript sub-path resolver
+    return (lookupArgs: ILookupProps) => {
```

Comment:
Just to be clear this tsResolver is a wrapper for `fileLookup` ?!

Reply:
Yes, but it injects its own `subPathResolver` instead of the default.

The reason this is done this way is that we want to use all of the logic inside fileLookup (e.g. look for exact match, then with extension, then index.js, etc.) but we want every one of those lookups to see if there is an associated input file and load that instead.

So for example, if someone `import 'mypkg'` then this will call into `fileLookup` and will read the `package.json` which has a `"main"` of `"dist/index.js"` and when it checks the subpath resolver to see if that exists, typescript can look for `"src/index.ts"` instead.

Reply:
Playground examples added.

There's `playground/ts-monorepo-1` which is Yarn 1, and `playground/ts-monorepo-2` which is Yarn 2.

Note that the Yarn 2 playground package might not work correctly due to the placement of the `playground` folder in the repo.  Yarn 2 can find the `package.json` of `fuse-box` which will cause it to complain.

Reply:
There is still one more problem to resolve here which can be demonstrated in the playground example.

If you run `playground/ts-monorepo-1` and make a change to `@myorg/bar/src/watchtrigger.ts`, you can see the watcher pick it up (which is good) but the change doesn't make it through cache/hmr.

This seems to be because `@myorg/bar` is not seen as a "user" package.  I have a (proposed) fix for this, which I'll submit in a separate PR for discussion.

Reply:
Awesome! I will check it tomorrow on Monday! There are files that do some simulations? Could you leave me instructions on what to run first?

Reply:
As I was writing some instructions, it occurred to me I could make some things easier so I am modifying the playground code to be easier to use.  Both playgrounds are basically the same, but `-2` is tweaked for yarn 2.  Both of them have READMEs in them that have some explanation and instructions.  I'm going to make some quick improvements to these, and update this PR in the next few minutes.

Also: I can't easily submit a separate PR for my proposed fix for the "user / non-user package issue" because it is based too heavily on changes in this PR and can't be separated.  So I'm going to put it back in this PR.  My change works and all tests pass, but I want to make sure it is heading in the right direction.

Reply:
@wagerfield I am going to start reviewing in in a few hours. If you need to continue with other changes, it's probably better if you base off of this branch. It is, however, a fork. That's why I was suggesting moving it to a fuse-box repo itself.